src/coreComponents/physicsSolvers/multiphysics/SinglePhasePoromechanicsKernel.hpp

```diff
@@ -135,189 +128,194 @@ class SinglePhase :
-      localResidualMass{ 0.0 },
-      dLocalResidualMass_dDisplacement{ {0.0} },
-      dLocalResidualMass_dPressure{ {0.0} },
-      localFlowDofIndex{ 0 }
+      Base::StackVariables()
     {}
 
-#if !defined(CALC_FEM_SHAPE_IN_KERNEL)
-    /// Dummy
-    int xLocal;
-#else
-    /// C-array stack storage for element local the nodal positions.
-    real64 xLocal[numNodesPerElem][3];
-#endif
-
-    /// Stack storage for the element local nodal displacement
-    real64 u_local[numNodesPerElem][numDofPerTrialSupportPoint];
-
-    /// Stack storage for the element local nodal incremental displacement
-    real64 uhat_local[numNodesPerElem][numDofPerTrialSupportPoint];
+    /// Mass accumulation
+    real64 fluidMassIncrement{};
+    /// Derivative of mass accumulation wrt volumetric strain increment
+    real64 dFluidMassIncrement_dVolStrainIncrement{};
+    /// Derivative of mass accumulation wrt pressure
+    real64 dFluidMassIncrement_dPressure{};
 
-    real64 ( &localResidualMomentum )[numDispDofPerElem];
-    real64 ( &dLocalResidualMomentum_dDisplacement )[numDispDofPerElem][numDispDofPerElem];
-    real64 dLocalResidualMomentum_dPressure[numDispDofPerElem][1];
+    // Storage for mass residual and degrees of freedom
 
-    real64 localResidualMass[1];
-    real64 dLocalResidualMass_dDisplacement[1][numDispDofPerElem];
-    real64 dLocalResidualMass_dPressure[1][1];
-
-    /// C-array storage for the element local row degrees of freedom.
-    globalIndex localFlowDofIndex[1];
+    /// Mass balance residual
+    real64 localResidualMass[1]{};
+    /// Derivative of mass balance residual wrt displacement
+    real64 dLocalResidualMass_dDisplacement[1][Base::StackVariables::numDispDofPerElem]{};
+    /// Derivative of mass balance residual wrt pressure
+    real64 dLocalResidualMass_dPressure[1][1]{};
 
   };
   //*****************************************************************************
 
   /**
-   * @brief Copy global values from primary field to a local stack array.
-   * @copydoc ::geosx::finiteElement::ImplicitKernelBase::setup
-   *
-   * For the SinglePhase implementation, global values from the displacement,
-   * incremental displacement, and degree of freedom numbers are placed into
-   * element local stack storage.
+   * @brief Helper function to compute 1) the total stress, 2) the body force term, and 3) the fluidMassIncrement
+   * using quantities returned by the PorousSolid constitutive model.
+   * This function also computes the derivatives of these three quantities wrt primary variables
+   * @param[in] k the element index
+   * @param[in] q the quadrature point index
+   * @param[in] strainIncrement the strain increment used in total stress and porosity computation
+   * @param[inout] stack the stack variables
    */
   GEOSX_HOST_DEVICE
   GEOSX_FORCE_INLINE
-  void setup( localIndex const k,
-              StackVariables & stack ) const
+  void
+  smallStrainUpdate( localIndex const k,
```

Comment:
same thing for this function. I think it can be moved to the base class.

Reply:
Yes, for both, we can move them to the base class. There is one hurdle that I discovered while working on the PR: if we want the base class function `PoromechanicsBaseKernel::smallStrainUpdate` to call the derived class function `SinglePoromechanicsKernel::computeFluidIncrement`, the signature of the base class function has to look like:
```
  template< typename LEAF >
  GEOSX_HOST_DEVICE
  GEOSX_FORCE_INLINE
  void
  smallStrainUpdate( localIndex const k,
                     localIndex const q,
                     real64 const ( &strainIncrement )[6],
                     typename LEAF::StackVariables & stack ) const
  { ... }
```
where `LEAF` is `SinglePhasePoromechanicsKernel< SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE >` or  `MultiphasePoromechanicsKernel< SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE >`, so it scared me a little bit :). I don't know exactly what the implications of all these templates in such a simple function are on compilation time for instance (maybe none?). I will do some tests on Travis.

Reply:
It is more difficult than I anticipated. If I move this function to the base kernel class, in the addition to the `LEAF` template mentioned above, I have to define dummy implementations of `computeFluidIncrement` and `computeBodyForce`, and it it really tricky which function is actually called in the code. So for now, `quadraturePointKernel` and `smallStrainUpdate` are duplicated. I think it is ok because these functions are very simple.

Reply:
I wish we could just use virtual methods....